### PR TITLE
Select SuggestionItem if there's no match in the list

### PR DESCRIPTION
**Pull request by genlu**
Fix https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1423475

When we are in suggestion-mode (either by toggling suggestion mode or in naming context,) select the SuggestionItem if there's no CompletionItem in the list that matches the filter text in anyway.

Note the change of selection box in the screen recording below
SelectSuggestion

### CODE REVIEW

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs [hunk 1]
```diff
@@ -670,8 +682,8 @@ private static bool IsPotentialFilterCharacter(char c)
 
             private ItemSelection UpdateSelectionBasedOnSuggestedDefaults(IReadOnlyList<MatchResult<VSCompletionItem>> items, ItemSelection itemSelection, CancellationToken cancellationToken)
             {
-                // Editor doesn't provide us a list of "default" items.
-                if (_snapshotData.Defaults.IsDefaultOrEmpty)
+                // Editor doesn't provide us a list of "default" items, or we select SuggestionItem (because we are in suggestion mode and have no match in the list)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs:685]: might be worthwhile to explain that we represent the SuggestionItem with a negative number.

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs [hunk 2]
```diff
@@ -670,8 +682,8 @@ private static bool IsPotentialFilterCharacter(char c)
...
+                if (_snapshotData.Defaults.IsDefaultOrEmpty || itemSelection.SelectedItemIndex < 0)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs:686]: my personal preference would be if we defined a constant for this.  e.g. only -1 should match, not any negative number.

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs [hunk 3]
```diff
@@ -266,6 +266,11 @@ private void AddCompletionItems(List<MatchResult<VSCompletionItem>> list, Cancel
                     MatchResult<VSCompletionItem> bestOrFirstMatchResult;
                     if (filteredItemsBuilder.Count == 0)
                     {
+                        // When we are in suggestion mode and there's nothing in the list matches what user has typed in any ways,
+                        // we should select the SuggestionItem instead.
+                        if (_snapshotData.DisplaySuggestionItem)
+                            return new ItemSelection(SelectedItemIndex: -1, SelectionHint: UpdateSelectionHint.SoftSelected, UniqueItem: null);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs:272]: we shou dintroduce a constant for this -1, the repetition is not great.

Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs]: done

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 4]
```diff
@@ -5975,16 +5975,19 @@ class C
...
-            case identifier $$
+            case identifier$$
         }
     }
 }
                               </Document>,
                   showCompletionInArgumentLists:=showCompletionInArgumentLists)
 
-                state.SendTypeChars("z")
+                state.SendTypeChars(" ")
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb:5985]: i would perfer not changing what the test *does* only what it checks.  if we want to test a different set of user behaviors, we shoudl add a new test.

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs [hunk 5]
```diff
@@ -48,6 +51,9 @@ private sealed class CompletionListUpdater
 
             private readonly Action<IReadOnlyList<(RoslynCompletionItem, PatternMatch?)>, string, IList<RoslynCompletionItem>> _filterMethod;
 
+            private bool ShouldSelectSuggestionItem
+                => _snapshotData.DisplaySuggestionItem && _filterText.Length > 0;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs:55]: not sure i understand the concept here.  We want to select the suggestion item whnever we're displaying it, and there is some filter text?  that seems odd to me.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs:55]: perhaps ShouldSelectSuggestionItemWhenAllOtherItemsFilteredOut ?

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb (hunk 4 above)
Comment (genlu) [reply src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb:5985]: I agree with this in general, but in this particular case, I'd argue that the changed behavior still reflects the original intention of the test (based on the test name at least)

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs (hunk 5 above)
Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs:55]: > ShouldSelectSuggestionItemWhenAllOtherItemsFilteredOut

Yup, that makes more sense

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb:5985]: the test name can be updated to now reflect what it was previously doing.  you can then use thsi test name to create a new test that does something different.  Tests are good, even if they're testing something else, and i'd prefer we never get rid of an existing test unless there is a very very good reason to do so (e.g. the component they test doesn't even exist anymore).

Comment (genlu) [reply src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: OK. In that case, there's probably no point to cove the new behavior here, which is covered in a separate test. Reverted this to its original behavior.

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**